src/raft/raft.go

```diff
@@ -326,33 +311,45 @@ func (rf *Raft) HeartBeat(args *HeartBeatRequest, reply *HeartBeatReply) {
 			for i, v := range rf.log[args.PrevLogIndex+1 : rf.lastLogIndex()+1] {
 				if i < len(args.Entries) && v.Term != args.Entries[i].Term {
 					rf.log = rf.log[:args.PrevLogIndex+1+i] // here use Index Not Term!...
+					rf.persist()                            // at HeartBeat log updates
+					// SUBTLE BUG PLACE: when cutting rf.log, rf.nextIndexes[x] needs to be cut as well.
+					// This is not really needed for now as when this cut happens the node is already a follower
+					// Then rf.nextIndexes should not be used anymore, and it always gets updated when it steps up as a leader
 					break
 				}
 			}
+
 			// [PAPER] AppendEntries-RPC - Receiver implementation: 4. Append any new Entries not already in the log
 			// leader:       [x, 1, 2, 3, 4] --- args.PrevLogIndex->2, Entries->[3,4]
 			// follower now: [x, 1, 2, 3 ] --- rf.lastLogIndex()->3, diff->1
 			diff := rf.lastLogIndex() - args.PrevLogIndex
-			if diff > 0 {
-				fmt.Printf("aaaaaaaaaaaaaaaaaaaaaaa entries has some duplicates append, diff:%v log len:%v\n", diff, len(args.Entries[diff:]))
+
+			if diff >= len(args.Entries) {
+				// for case like:
+				// log: [0, 1, 2, 3, 4, 5, 6] -------------------------> PrevLogIndex=2,
+				//               [3, 4,] ------------------------------> len(args.Entries)=2
+				//         diff := rf.lastLogIndex() - args.PrevLogIndex = (7-2) = 5 > 2 will cause issue
+				// So if after the above "cutting" step results this happens, we just do nothing as the logs has all the same element already.
+				// Here could introduce a subtle bug as in the above example entry [5, 6] might be some old term in valid entry which will be cleaned later
+				// And here we do nothing, but rf.commitIndex could be updated to the very end and could cusing
+				// So we do this `newCommitIndex := min(args.LeaderCommit, args.PrevLogIndex+len(args.Entries))` below
+				// instead of `newCommitIndex := min(args.LeaderCommit, rf.lastLogIndex())`, otherwise, TestFigure8Unreliable2C will give commits aligned error under critical conditions
+			} else {
+				if diff > 0 {
+					fmt.Printf("aaaaaaaaaaaaaaaaaaaaaaa entries has some duplicates append, diff:%v log len:%v\n", diff, len(args.Entries[diff:]))
+				}
+				rf.log = append(rf.log, args.Entries[diff:]...)
+				rf.persist() // at HeartBeat log updates
 			}
-			rf.log = append(rf.log, args.Entries[diff:]...)
 		}
 		// [PAPER] AppendEntries-RPC - Receiver implementation: 5.  If LeaderCommit > commitIndex, set commitIndex = min(LeaderCommit, index of last new entry)
 		// Follower commit
 		if args.LeaderCommit > rf.commitIndex {
-			newCommitIndex := min(args.LeaderCommit, rf.lastLogIndex())
+			// IMPORTANT - SUBTLE - To allow TestFigure8Unreliable2C give no commit order issue, need to use args.PrevLogIndex+len(args.Entries), see comments above.
+			newCommitIndex := min(args.LeaderCommit, args.PrevLogIndex+len(args.Entries))
```

Comment:
Important fix here - as we added some optimization on 
1. Leader only sends part of new entries when there are more than a max number (500 in this case) of new entries that are not on the follower
2. When a Follower is handling the heartbeat, it does not touch the log if the incoming entries have been existing/duplicated, leaving some entries after the incoming entry part untouched. However, the entries in the latter part might not be valid.

So with those optimizations mentioned above, at the place when we update the follower's `rf.commitIndex`, we cannot go as far as the leader's `commitIndex`, but only goes to the end index of the incoming replicating entry parts - which is `args.PrevLogIndex+len(args.Entries)`